docs/core/whats-new/dotnet-core-3-0.md

```diff
@@ -0,0 +1,496 @@
+                if (!json.TryGetInt32Value(out int valueInteger))
+                {
+                    throw new FormatException();
+                }
+                break;
+            case JsonTokenType.True:
+            case JsonTokenType.False:
+                bool valueBool = json.GetBooleanValue();
+                break;
+            case JsonTokenType.Null:
+                break;
+            default:
+                throw new ArgumentException();
+        }
+    }
+
+    dataUtf8 = dataUtf8.Slice((int)json.BytesConsumed);
+    JsonReaderState state = json.CurrentState;
+}
+```
+
+The .NET ecosystem has relied on [Json.NET](https://www.newtonsoft.com/json) and other popular JSON libraries, which continue to be good choices. JSON.NET uses .NET strings as its base datatype, which are UTF-16 under the hood. 
+
+In .NET Core 2.1 and 3.0, we added new APIs that makes it possible to write JSON APIs (such as `Utf8JsonReader`) that require much less memory, based on using `Span<T>` and UTF-8 strings, and better serve the needs of high-throughput applications like Kestrel, ASP.NET Core web server.
+
+## Ranges and indices
+
+The new `Index` type can be used for indexing. You can create one from an `int` that counts from the beginning, or with a prefix `^` operator (C#) that counts from the end:
+
+```csharp
+Index i1 = 3;  // number 3 from beginning
+Index i2 = ^4; // number 4 from end
+int[] a = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
+Console.WriteLine($"{a[i1]}, {a[i2]}"); // "3, 6"
+```
+
+There is also a `Range` type, which consists of two `Index` values, one for the start and one for the end, and can be written with a `x..y` range expression (C#). You can then index with a `Range` in order to produce a slice:
+
+```csharp
+var slice = a[i1..i2]; // { 3, 4, 5 }
+```
+
+> [!NOTE]
+> Only [C# 8.0](https://blogs.msdn.microsoft.com/dotnet/2018/11/12/building-c-8-0/) supports syntax for `Range` and `Index`.
+
+## Async streams
+
+The `IAsyncEnumerable<T>` type is a new asynchronous version of `IEnumerable<T>`. The language lets you `await foreach` over these to consume their elements, and `yield return` to them to produce elements.
+
+The following example demonstrates both production and consumption of async streams. The `foreach` statement is async and itself uses `yield return` to produce an async stream for callers. This pattern (using `yield return`) is the recommended model for producing async streams.
+
+```csharp
+async IAsyncEnumerable<int> GetBigResultsAsync()
+{
+    await foreach (var result in GetResultsAsync())
+    {
+        if (result > 20) yield return result; 
+    }
+}
+```
+
+In addition to being able to `await foreach`, you can also create async iterators, for example, an iterator that returns an `IAsyncEnumerable/IAsyncEnumerator` that you can both `await` and `yield` in. For objects that need to be disposed, you can use `IAsyncDisposable`, which various BCL types implement, such as `Stream` and `Timer`.
```

Comment:
We should mention dotnet/Roslyn#31268 because the example above won't work correctly in Core.

Reply:
Should I not include it then?

Reply:
I'd leave it in. Consuming async streams works fine. Generating them inside a `foreach` loop generates code that only enumerates the first item. The workaround is to use a `for` loop.